hkmc2/shared/src/main/scala/hkmc2/semantics/Specialiser.scala

```diff
@@ -0,0 +1,753 @@
+                  case _: UnitLit => S(tUnit)
+                  case _ => N
+                case Fld(_, Ref(r), _) => ctx.get(r.nme).flatMap(_.asInstanceOf[Binding].typ)
+                case _ => N
+              case _ => N
+
+            log(s"Found application of ${name} with type ${typ.map(_.tree.name).getOrElse("error")}")
+            val newApps = typ match
+              case Some(t) => apps + (name -> (t :: apps.getOrElse(name, Nil).filterNot(_ == t)))
+              case N => apps
+
+            // val newName: Ident = Ident(name + "_" + typ.map(_.tree.name).getOrElse("oops"))
+            // (specApp, newApps)
+            (t, apps)
+          case _: Sel => (t, apps)
+          case _ =>
+            raise(ErrorReport(msg"I messed up :(" -> t.toLoc :: Nil)) // FIXME
+            (t, apps)
+      case il @ IfLike(_, desug) => desug match
+        case Let(s, b, t) =>
+          val (nb, lowerApps) = term(b, apps)
+          (il.copy(desugared = Let(s, nb, t))(il.normalized), lowerApps)
+        case Else(d) =>
+          val (nd, lowerApps) = term(d, apps)
+          (il.copy(desugared = Else(nd))(il.normalized), lowerApps)
+        case _ => (il, apps)
+      case Lam(params, body) => // TODO: specialise the lambda
+        val (newBody, newApps) = term(body, apps)
+        (Lam(params, newBody), newApps)
+      case Forall(tvs, outer, body) => // FIXME
+        val (newTerm, newApps) = term(body, apps)
+        (Forall(tvs, outer, newTerm), newApps)
+      case Quoted(b) =>
+        val (newTerm, newApps) = term(b, apps)
+        (Quoted(newTerm), newApps)
+      case Unquoted(b) =>
+        val (newTerm, newApps) = term(b, apps)
+        (Unquoted(newTerm), newApps)
+      case Region(name, body) =>
+        val (newTerm, newApps) = term(body, apps)
+        (Region(name, newTerm), newApps)
+      case Deref(ref) =>
+        val (newTerm, newApps) = term(ref, apps)
+        (Deref(newTerm), newApps)
+      case Ret(expr) =>
+        val (newTerm, newApps) = term(expr, apps)
+        (Ret(newTerm), newApps)
+      case Throw(expr) =>
+        val (newTerm, newApps) = term(expr, apps)
+        (Throw(newTerm), newApps)
+      case Try(body, finallyDo) =>
+        val (b1, apps1) = term(body, apps)
+        val (b2, apps2) = term(finallyDo, apps1)
+        (Try(b1, b2), apps2)
+      case b: Blk => block(b, apps)
+      case _ => (t, apps) // TODO: Handle the few other term types
+
+  def topLevel(b: Blk): Blk = b
+    // block(b, Map.empty)(using Ctx.empty)._1
+    //
+end Specialiser
+
+import scala.collection.mutable
+import mlscript.utils.*, shorthands.*
+import syntax.*
+import semantics.*
+
+class SimpleSub(val tl: TraceLogger):
+  import tl.*
+
+  enum SimpleType:
+    case Variable(state: VariableState)
+    case Primitive(name: Str)
+    case Function(lhs: SimpleType, rhs: SimpleType)
+    case Record(fields: Ls[(Str, SimpleType)])
+    case ClassType(sym: ClassSymbol, supers: Ls[SimpleType], members: Map[Str, SimpleType])
```

Comment:
This is more like `ClassInfo` (which is _not_ a `SimpleType`). The types of actual class instances should be a different constructor (which points back to the `ClassInfo`).